Support declarative partitioning

fljdin opened the pull request:
Enhances API with "partitions" and "subpartitions" new tables. Add steps in materialize_foreign_table() to support partitioned tables and partitions creation when needed. See README.md for further description.

Supports LIST (with default partition), RANGE, HASH and composite partitioning. The last position of a HASH partition is read before partition creation to obtain modulus value. Positions must start at 1.

fljdin committed:
Support declarative partitioning

Enhances API with "partitions" and "subpartitions" new tables. Add steps in
materialize_foreign_table() to support partitioned tables and partitions creation
when needed. See README.md for further description.

Supports LIST (with default partition), RANGE, HASH and composite partitioning.
The last position of a HASH partition is read before partition creation to
obtain modulus value. Positions must start at 1.

laurenz commented on db_migrator--0.9.0.sql:
```diff
@@ -20,25 +20,179 @@ CREATE FUNCTION materialize_foreign_table(
+                  WHEN v_type = 'HASH' THEN format(
+                     'FOR VALUES WITH (modulus %s, remainder %s)', 
+                     v_lastpos, v_pos - 1
+                  )
```
Perhaps an error message for other values...

fljdin replied:
`v_type` comes from `partitions.partitions` column, already rules by `CHECK` constraint.

```sql
CONSTRAINT partition_chk_type 
         CHECK (type IN ('LIST', 'RANGE', 'HASH'))
```

laurenz replied:
Yes, that is certainly good enough; my oversight.

laurenz commented on db_migrator--0.9.0.sql:
```diff
@@ -20,25 +20,179 @@ CREATE FUNCTION materialize_foreign_table(
+                        WHEN v_type = 'HASH' THEN format(
+                           'FOR VALUES WITH (modulus %s, remainder %s)', 
+                           v_lastpos, v_pos - 1
+                        )
```
Again, an error message for other value would be good.

fljdin replied:
Same here, I supposed constraints do the job

```sql
CONSTRAINT subpartition_chk_type 
         CHECK (type IN ('LIST', 'RANGE', 'HASH'))
```

laurenz replied:
Absolutely. I didn't see the constraint.

laurenz commented on README.md:
```diff
@@ -337,6 +337,64 @@ are done with the migration.
+  partitioning
+
+- `position` (type `integer`): defines the order of the table partitions (must
+  start at 1)
```
Do we really need that? Partitions don't have an order.
I know you use it to determine the number of hash partitions, but that could just as well be determined by the row count of the result set.
You could `MOVE` to the end and
```
GET DIAGNOSTICS integer_var = ROW_COUNT;
```

fljdin replied:
`position` column has been deleted as discussed in #12 

See https://github.com/fljdin/db_migrator/commit/209a2e12b15bf9082ac4c8e28cdb9226829e9883

laurenz reviewed (commented):
This is just a code review; I didn't test it yet.

Thanks for the contribution! This is promising!

laurenz commented on db_migrator--0.9.0.sql:
```diff
@@ -20,25 +20,179 @@ CREATE FUNCTION materialize_foreign_table(
+                  WHEN v_default THEN 'DEFAULT'
+                  WHEN v_type = 'LIST' THEN format(
+                     'FOR VALUES IN (%s)', 
+                     v_values
```
There should be a call to `translate_expression_fun` here and in the other branches.
The same applies to subpartitioning.

fljdin replied:
As wrote earlier, I've decided to use translate expression in plugin side because of array construct. Could be enhanced in the future.

laurenz replied:
Hm. That violates the design principle that the foreign tables should show the original data from the source, and all translations should happen in `db_migrate_refresh()`.

Please add that in the code after line 559 of `db_migrator--0.9.0.sql` instead.

fljdin replied:
Finally did it! https://github.com/cybertec-postgresql/db_migrator/pull/11/commits/e0eab88da6e0289233e4bdb2f30794975f740b6f

No regression with ora_migrator, especially expression boundaries defined in the subpartitions test

 https://github.com/cybertec-postgresql/ora_migrator/pull/42/commits/7b3cd03fa4955596da056902e0a9db47e215c161

laurenz replied:
Thanks!

laurenz reviewed (commented):
One more think I noticed after reading the `ora_migrator` patch.

laurenz commented on README.md:
```diff
@@ -337,6 +337,64 @@ are done with the migration.
 
+### `partitions` ###
+
+- `schema` (type `name`): schema of the table with the partition
```
How about "schema of the partitioned table"?

fljdin replied:
Changed in 6b6497ebd7036d27fac47843743abc787086c51c

laurenz replied:
&#x2713;

laurenz commented on README.md:
```diff
@@ -337,6 +337,64 @@ are done with the migration.
+
+- `schema` (type `name`): schema of the table with the partition
+
+- `table_name` (type `name`): name of the related table
```
Similarly, I'd prefer "name of the partitioned table".

fljdin replied:
Changed in 6b6497ebd7036d27fac47843743abc787086c51c

laurenz commented on README.md:
```diff
@@ -337,6 +337,64 @@ are done with the migration.
+  start at 1)
+
+- `values` (type `text`): for a `RANGE` method, contains the upper bound value,
+  for a `LIST` method, contains a list of comma-separated values
```
Only the upper bound for range partitioning? Why not both?
Perhaps that has to do with Oracle's metadata, but I think it is odd and undesirable in general.
At least now I understand why you have the `position` that I criticized before.
Are Oracle partition boundaries inclusive or exclusive (I am too lazy to check right now)?

fljdin replied:
In Oracle and MySQL, it seems to be a catalog pratice to store upper bound only, see documentation below

* ([Oracle](https://docs.oracle.com/en/database/oracle/oracle-database/19/refrn/ALL_TAB_PARTITIONS.html)) ALL_TAB_PARTITIONS.HIGH_VALUE: Partition bound value expression
* ([MySQL](https://dev.mysql.com/doc/refman/8.0/en/information-schema-partitions-table.html)) INFORMATION_SCHEMA.PARTITIONS.PARTITION_DESCRIPTION: For a RANGE partition, it contains the value set in the partition's VALUES LESS THAN clause, which can be either an integer or MAXVALUE.

The upper bound is excluded

* ([Oracle](https://docs.oracle.com/en/database/oracle/oracle-database/21/vldbg/partition-concepts.html#GUID-D6CC12F9-81A5-4E89-872F-024C6161C0E8)) Each partition has a VALUES LESS THAN clause, that specifies a non-inclusive upper bound for the partitions.
* ([MySQL](https://dev.mysql.com/doc/mysql-partitioning-excerpt/8.0/en/partitioning-range.html)) MySQL must be able to evaluate the expression's return value as part of a LESS THAN (<) comparison.

fljdin replied:
With other DBMS, looks like more tedious

* ([SQL Server](https://learn.microsoft.com/en-us/sql/t-sql/statements/create-partition-scheme-transact-sql?view=sql-server-ver16#a-creating-a-partition-scheme-that-maps-each-partition-to-a-different-filegroup)) Bound are build from a list of values where upper bound is included if `RANGE LEFT` is used in creation statement
* ([Sybase](https://infocenter.sybase.com/help/index.jsp?topic=/com.sybase.infocenter.dc32300.1600/doc/html/san1390612300205.html)) Every partition has an inclusive upper bound, which is specified by the values <= clause when the partition is created. 
* ([DB2](https://www.ibm.com/docs/en/db2/11.5?topic=tables-defining-ranges)) INCLUSIVE indicates that all values equal to the specified value are to be included in the data partition containing this boundary.

laurenz replied:
Weird. I checked with list partitioning, and Oracle has the whole list in `HIGH_VALUE`.

I still think that the API should be general, even if both Oracle and MySQL have no support for gaps in the partitioned space. What if someone wants to write a plugin for a database that supports that?
Also, having the lower bound would also make it easier to exclude single partitions from the migration.

laurenz replied:
Thanks for the research! We won't be able to translate inclusive upper partition bounds to PostgreSQL.
But I think we don't have to worry about that. People will just have to live with some rows ending up in a different partition.

fljdin replied:
Hope #12 will handle all of newcoming plugins!

laurenz commented on README.md:
```diff
@@ -337,6 +337,64 @@ are done with the migration.
+  partition 
+
+- `migrate` (type `boolean`, default `TRUE`): `TRUE` if the constraint should
+  be migrated
```
This looks like it was copied from the constraints; should be "partition".

laurenz replied:
The way you handle range partition boundaries right now, that would mean that setting `migrate` to FALSE for a partition would extend the boundaries of the previous one.
Is that intended? I think it is surprising and not what I would expect.

fljdin replied:
Good catch, especially for `RANGE` partitioning. 

Another way to transform a partitioned table as a simple table could be simply remove all partition specifications from `plsql_stage.partitions` related to source table. But, problem still there if only one partition is deleted from scheme.

laurenz replied:
I just realized that there is another problem.

When the data are migrated, you just select from the partitioned table, that is, you get all data from all partitions, right? That would fail to load on PostgreSQL if you omitted one partition.

I am not sure what to do about that. I have two ideas:

1. Change the way data are migrated so that it is partition by partition (that is the hardest solution).
2. Only provide the option to migrate partitioning with all partitions as they are or not migrate partitioning at all. That would be good enough from my point of view.
   Then we should do away with the `migrate` column for partitions. We can either introduce a `migrate_partitioning` flag on the table or expect that the user deletes the partitions from the migrator table.

laurenz commented on README.md:
```diff
@@ -337,6 +337,64 @@ are done with the migration.
+  subpartition 
+
+- `migrate` (type `boolean`, default `TRUE`): `TRUE` if the constraint should
+  be migrated
```
Again "partition", not "constraint".

laurenz commented on README.md:
```diff
@@ -337,6 +337,64 @@ are done with the migration.
+
+- `schema` (type `name`): schema of the table with the partition
+
+- `table_name` (type `name`): name of the related table
```
My comments to the `partitions` table apply here too.

laurenz commented on README.md:
```diff
@@ -337,6 +337,64 @@ are done with the migration.
+
+- `partition_name` (type `name`): name of the parent partition
+
+- `subpartition_name` (type `name`): name of the subpartitions
```
The plural "subpartition**s**" is wrong.

laurenz commented:
I thought some more about partition boundaries.
Perhaps they can be passed as `text[]`: one value for HASH, two for RANGE and arbitrarily many for LIST.
Then there is the problem of numbers vs. other data types. Perhaps we can require that the values in the `text[]` are already properly quoted, so that we can simply use them as they are.

fljdin committed:
Update README.md
Reviewed by laurenz

fljdin committed:
Define lower and upper boundaries for partitioning

partitions.values and subpartitions.valuess must be arrays depending on
partition type Removes position and migrate columns.

Related to #12

fljdin committed:
Translates boundaries expressions on refresh

Use of v_translate_expression inside partitions and subpartitions refresh step.
As values represents an array of specification, it loops over unnested values to
perform translation before reconstruct array.

laurenz commented:
Ok, I think I have tortured you enough with my complaints, and thanks for your patience.
I will take it from here and get this merged.

fljdin commented:
Good news! All of your suggestions sounded good to me, thanks for your feedback!

laurenz commented:
Thanks again.
